// data-marauder/lib/myrob/robot.js
/**
 * author:nero
 * version:v1.0
 * robot will help me to control all things
 */
'use strict'
const Product = require('../../../../midserver/admin/service/product')
let product = new Product()
const Topic = require('../../../../midserver/admin/service/topic')
let topic = new Topic()
const Basemysql = require('../../../../midserver/admin/base/mysql')
let basemysql = new Basemysql()

const Excel = require('./excel')
let excel = new Excel()
let Delay = require('keeper-core/lib/delay')
let delay = new Delay()
let crypto = require('crypto')

// topic id for this time
let topicID

// constructor
class Robot {
  constructor() {
    this.options = {
      stream: process.stdout
    }
  }

  async robot() {
    console.log('robot is running!')
    let datalist = await this.excel()
    await delay.delay(1, true)
    topicID = await topic.getTopicIdByName(datalist.filename)
    console.log(topicID, datalist.filename)
    if (topicID) {
      this.proxycontrol(datalist.excel)
    } else {
      console.log('topic id is empty!'.red)
    }
  }

  async excel() { // transfrom excel data to json
    let datalist = await excel.boot()
    // console.log(datalist)
    console.log('excel data is read complete! proxy will working at 2 seconds later!')
    return datalist
  }

  // init the plugin 'phantom'
  async mysql(datalist, index, proid) {
    proid = proid + parseInt(index)
    return new Promise(async (resolve) => {
      let md5 = crypto.createHash('md5')
      let codeid = md5.update(proid.toString()).digest('hex')
      codeid = codeid.substr(22)
      /** A proid
       *  B proname
       *  C mainimg
       *  F price
       *  L prohref
       *  rowkey: ['A', 'B', 'C', 'F', 'G', 'L', 'M', 'P', 'S']
       */
      let data = {
        pro_id: datalist.A,  // 商品ID
        name: datalist.B,  // 商品名称
        code: codeid,  // 商品代码
        main_img: datalist.C,  // 主图
        sell_price: datalist.F,  // 售价
        sales_volume: datalist.G,  // 月销售量
        href: datalist.L,  // 购买链接
        taoword: datalist.M,  // 淘口令
        coupon: datalist.P,  // 优惠券
        couponhref: datalist.S, // 优惠券链接
        // pro_kind: datalist.W, // mysort
        topic_id: topicID.id  // 专题ID
      }
      // add product and get infor after insert.
      let haspro = await product.hasPro(data.pro_id)
      if (haspro) {
        await product.updatePro(data)
      } else {
        await product.addPro(data)
      }
      resolve()
    })
  }

  async proxycontrol(datalist) { // control of proxy and mysql
    let proid = await product.getProID()
    let count
    for (let i in datalist) {
      await this.mysql(datalist[i], i, proid.id)
      count = parseInt(i) + 1
    }
    console.log('count : ' + count)
    console.log('all mission is finished!')
    basemysql.endconn()
  }
}

module.exports = Robot
