// keeper-core/lib/infor.js
/**
 * Created by nero on 2019/1/22.
 * get user infor
 */
const inquirer = require('inquirer');
let fs = require('fs')
let path = require('path')
const Initnpm = require('keeper-core/lib/npm')
let mynpm = new Initnpm()
const Writefile = require('keeper-core/lib/writefile')
let writefile = new Writefile()

class infor {
  async config() {
    return new Promise(async (resolve) => {
      console.log('请填写项目信息：')
      const promptList = [{
        type: 'input',
        message: '设置一个用户名:',
        name: 'name',
        default: "test_user" // 默认值
      }, {
        type: 'list',
        message: '请选择一种水果:',
        name: 'fruit',
        choices: [
          "Apple",
          "Pear",
          "Banana"
        ],
        filter: function (val) { // 使用filter将回答变为小写
          return val.toLowerCase();
        }
      }, {
        type: "checkbox",
        message: "选择颜色:",
        name: "color",
        choices: [
          {
            name: "red"
          },
          new inquirer.Separator(), // 添加分隔符
          {
            name: "blur",
            checked: true // 默认选中
          },
          {
            name: "green"
          },
          new inquirer.Separator("--- 分隔符 ---"), // 自定义分隔符
          {
            name: "yellow"
          }
        ]
      }];

      let result = await inquirer.prompt(promptList)
      resolve(result)
    })
  }

  checkplugin(plugin, readyconfig) {
    this.initpro(readyconfig)
    mynpm.init(readyconfig.pluginlist, plugin)
  }

  initpro(readyconfig) {
    let initfile = readyconfig.initfile

    for (let i in initfile) {
      if (!fs.existsSync(initfile[i].in)) {
        let inconf = path.join(__dirname, '/../tpl/system/' + initfile[i].out)
        let outconf = initfile[i].in
        writefile.copy(inconf, outconf)
      }
    }
  }

  bootstrap(plugin, readyconfig) {
    const promptList = [{
      type: 'Confirm',
      message: '检测到新项目，请确认当前环境为根目录:[y/n]',
      name: 'boot',
      default: 'y'
    }];

    inquirer.prompt(promptList).then(a => {
      if (a.boot === 'y') {
        this.checkplugin(plugin, readyconfig)
      } else if (a.boot !== 'n') {
        console.log('输入无效！')
      }
    })
  }

  boot(plugname, readyconfig, fn) {
    // check program running environment.
    let sysconf = path.resolve('./node_modules/' + plugname + '/config/sysconf.js')
    let isready = fs.existsSync(sysconf)
    if (isready) {
      let tempver = fs.readFileSync(path.join(__dirname, '/../../../package.json')).toString()
      let currversion = JSON.parse(tempver).version
      // console.log(plugname + ' : '.green + currversion.green)
      let preversion = require(sysconf)
      if (preversion.check_env === currversion) {
        fn()
      } else {
        this.bootstrap(plugname, readyconfig)
      }
    } else {
      this.bootstrap(plugname, readyconfig)
    }
  }
}

module.exports = infor
